---
title: "EpiCompare Report"
author: "Sera Choi"
output: 
  html_document: 
    toc: yes
    toc_float: yes
params:
  peakfile: 
    value: NULL
  output_dir: 
    value: NULL
  blacklist: 
    value: NULL
  picard_files:
    value : NULL
  reference: 
    value: NULL
  stat_plot:
    value: FALSE
  chrmHMM_plot:
    value: FALSE
  chipseeker_plot:
    value: FALSE
  enrichment_plot:
    value: FALSE
  save_output:
    value: FALSE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
# if save_output is TRUE, create a directory to save all outputs, if it doesn't exist already 
if(params$save_output){
  outfile_dir <- paste0(params$output_dir,"/EpiCompare_file")
  if(!dir.exists(outfile_dir)){
    dir.create(outfile_dir)
  }
}
# tidy input peakfiles to remove peaks in blacklisted regions and non-standard chromosomes 
# and check that the list is named, if not default filenames are used 
peaklist_tidy <- tidy_peakfile(peaklist = params$peakfile,
                               blacklist = params$blacklist)
# tidy input reference peakfile if provided
if (is.null(params$reference) == FALSE){
  reference_tidy <- tidy_peakfile(peaklist = params$reference,
                                  blacklist = params$blacklist)
}
```

EpiCompare compares different epigenetic datasets for quality control and benchmarking purposes. 
The report is divided into three sections: 

1. **General Metrics:** Metrics on fragments (duplication rate) and peaks (blacklisted peaks and peak widths) of input files
2. **Peak Overlap:** Percentage and statistical significance of overlapping and unon-overlapping peaks.
3. **Functional Annotation:** Functional annotation (ChromHMM, ChIPseeker and enrichment analysis) of peaks. 

<br>

##### Input Datasets 

* Total of `r length(peaklist_tidy)` peak files: 

```{r name, echo=FALSE}
# print peak file names and numerate 
n <- 1
for (name in names(peaklist_tidy)){
  print(paste0("File", n, ": ", name))
  n <- n + 1
}
```

* Reference peakfile: 

***

## 1. General Metrics {.tabset}

### Peak Information 

Column Description:

* PeakN before tidy: Total number of peaks including those blacklisted and those in non-standard chromosomes. 
* Blacklisted peaks removed (%): Percentage of blacklisted peaks present in the sample. [ENCODE blacklist](https://www.encodeproject.org/files/ENCFF000KJP/) includes regions in the hg19 genome that have 
anomalous and/or unstructured signals independent of the cell-line or experiment.
* Non-standard peaks removed (%): Percentage of peaks identified in non-standard and/or mitochondrial chromosomes.
Identified using `BRGenomics::tidyChromosomes()`.  
* PeakN after tidy: Total number of peaks after filtering blacklisted peaks and those in non-standard chromosomes. 

  NB: All analyses in EpiCompare are conducted on *tidied* datasets (i.e. blacklisted peaks and those in non-standard chromosomes removed)

<br>

```{r, echo=FALSE, warning=FALSE}
peak_info_df <- peak_info(peak_list = params$peakfile,
                          blacklist = params$blacklist)
# create table 
knitr::kable(peak_info_df)
# save output as table 
if(params$save_output){
  write.table(peak_info_df, file=paste0(outfile_dir,"/peak_info"), sep="\t")
}
```

***

### Fragment Information

Metrics on fragments is shown only if Picard summary is provided. See [manual](https://neurogenomics.github.io/EpiCompare/) for help. 

Column Description:

* Mapped_Fragments: Number of mapped read pairs in the file. 
* Duplication_Rate: Percentage of mapped sequence that is marked as duplicate.
* Unique_Fragments: Number of mapped sequence that is not marked as duplicate. 

<br>

```{r, echo=FALSE, warning=FALSE}
# only if picard_files is provided, run the function 
if (!is.null(params$picard_files)){
  fragment_info_df <- fragment_info(picard_list = params$picard_files)
  # create data frame 
  knitr::kable(fragment_info_df)
  # save output 
  if(params$save_output){
    write.table(fragment_info_df, file=paste0(outfile_dir,"/fragment_info"), sep="\t")
    }
}   
```

***

### Peak widths 

Distribution of peak widths in each sample. 

<br> 

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# create boxplot of peak widths 
width_plot <- width_boxplot(peaklist = peaklist_tidy)
print(width_plot)
# save boxplot output 
if(params$save_output){
  ggplot2::ggsave(filename = "width_plot.png",plot = width_plot, device = "png", path = outfile_dir)
  unlink("width_plot.png")
}
```

***

## 2. Peak Overlap {.tabset}

### Percentage Overlap  

Heatmap of percentage of overlapping peaks between samples. 
Hover over the heatmap for percentage values. 

<br>

```{r, echo=FALSE, fig.width= 9, warning=FALSE}
overlap_heatmap <- overlap_heatmap(peaklist = peaklist_tidy,
                  save_dir = NULL)
overlap_heatmap
# save output 
if(params$save_output){
  overlap_heatmap(peaklist = peaklist_tidy,
                  save_dir = paste0(outfile_dir,"/samples_percent_overlap.html"))
  unlink(paste0(outfile_dir,"/samples_percent_overlap_files"), recursive = TRUE)
  }

```

***

### Statistical Significance  

The plot is shown only if a reference peak file is provided and `stat_plot = TRUE`. 
Depending on the format of the reference file, EpiCompare outputs different plots: 

* Reference dataset has BED6+4 format (peakcalling performed with MACS2): EpiCompare generates paired boxplot 
per sample showing the distribution of -log10(q-value) of reference peaks that are overlapping and non-overlapping with the sample dataset. 
* Reference dataset does not have BED6+4 format: EpiCompare generates a barplot of percentage of overlapping sample peaks with the reference, coloured by statistical significance (adjusted p-value) of the overlap. 

Keys:
* Overlap: Sample peaks in Reference peaks 
* Unique: Sample peaks not in Reference peaks  

<br>

```{r, echo=FALSE, fig.width=9, warning=FALSE, message=FALSE}
# if reference is provided AND stat_plot=TRUE, run the function 
if (is.null(params$reference) == FALSE){
  if (params$stat_plot){
    stat_plot <- overlap_stat_plot(reference = reference_tidy,
                                   peaklist = peaklist_tidy)
    print(stat_plot[[1]])
    # save output 
    if(params$save_output){
      ggplot2::ggsave(filename = "stat_plot.png", plot = stat_plot[[1]], device = "png", path = outfile_dir)
      unlink("stat_plot.png")
      }
  }
}
```

***

## 3. Functional Annotation 

### 3.1 ChromHMM {.tabset}

[ChromHMM](http://compbio.mit.edu/ChromHMM/#:~:text=ChromHMM%20is%20software%20for%20learning%20and%20characterizing%20chromatin%20states.&text=ChromHMM%20is%20based%20on%20a,one%20or%20more%20cell%20types.) annotates and characterises peaks into different chromatin states.
The annotation used in EpiCompare were obtained from [here](http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeBroadHmm/).

#### All samples 

ChromHMM annotation of individual samples. 

```{r, echo=FALSE, fig.width=9, warning=FALSE, message=FALSE}
# if chrmHMM_plot=TRUE, create ChromHMM heatmaps 
if (params$chrmHMM_plot){
  # download chromHMM annotation files 
  chrHMM_url = "http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeBroadHmm/wgEncodeBroadHmmK562HMM.bed.gz"
  chrHMM = genomation::readBed(chrHMM_url)
  chrHMM_list = GenomicRanges::split(chrHMM, chrHMM$name, drop = TRUE) 
  # plot chromHMM
  samples_chrmHMM <- plot_chrmHMM(peaklist = peaklist_tidy,
                                  chrmHMM_annotation = chrHMM_list)
  print(samples_chrmHMM)
  # save output 
  if(params$save_output){
      ggplot2::ggsave(filename = "samples_chromHMM.png", plot = samples_chrmHMM, device = "png", path = outfile_dir)
      unlink("samples_chromHMM.png")
  }
}
```

#### **Overlap**: Sample peaks in Reference peaks 

ChromHMM annotation of sample peaks found in reference peaks.   

```{r, echo=FALSE, fig.width=9, warning=FALSE, message=FALSE}
if (params$chrmHMM_plot){
  if (is.null(params$reference)==FALSE){
    sample_in_ref_list <- list()
    # obtain overlapping peaks 
    for(file in peaklist_tidy){
      sample_in_ref <- IRanges::subsetByOverlaps(x = file, ranges = reference_tidy)
      sample_in_ref_list <- c(sample_in_ref_list, sample_in_ref)
    }
    names(sample_in_ref_list) <- names(peaklist_tidy)
    # annotate overlapping peaks with chromHMM annotations
    sample_in_ref_chrmHMM <- plot_chrmHMM(peaklist = sample_in_ref_list,
                                          chrmHMM_annotation = chrHMM_list)
    print(sample_in_ref_chrmHMM)
    # save output 
    if(params$save_output){
      ggplot2::ggsave(filename = "sample_in_ref_chromHMM.png", plot = sample_in_ref_chrmHMM, device = "png", path = outfile_dir)
      unlink("sample_in_ref_chromHMM.png")
    }
  }
}
```

#### **Overlap**: Reference peaks in Sample peaks 

ChromHMM annotation of reference peaks found in sample peaks.   

```{r, echo=FALSE, fig.width=9, warning=FALSE, message=FALSE}
if (params$chrmHMM_plot){
  if (is.null(params$reference)==FALSE){
    ref_in_sample_list <- list()
    # obtain overlapping peaks 
    for(file in peaklist_tidy){
      ref_in_sample <- IRanges::subsetByOverlaps(x = reference_tidy, ranges = file)
      ref_in_sample_list <- c(ref_in_sample_list, ref_in_sample)
    }
    names(ref_in_sample_list) <- names(peaklist_tidy)
    # annotate overlapping peaks with chromHMM annotations
    ref_in_sample_chrmHMM <- plot_chrmHMM(peaklist = ref_in_sample_list,
                                          chrmHMM_annotation = chrHMM_list)
    print(ref_in_sample_chrmHMM)
    # save output 
    if(params$save_output){
      ggplot2::ggsave(filename = "ref_in_sample_chromHMM.png", plot = ref_in_sample_chrmHMM, device = "png", path = outfile_dir)
      unlink("ref_in_sample_chromHMM.png")
    }
  }
}
```

#### **Unique**: Sample peaks not in Reference peaks 

ChromHMM annotation of sample peaks not found in reference peaks. 


```{r, echo=FALSE, fig.width=9, warning=FALSE, message=FALSE}
if (params$chrmHMM_plot){
  if (is.null(params$reference)==FALSE){
      sample_not_in_ref_list <- list()
      # obtain unique peaks 
      for(file in peaklist_tidy){
        sample_not_in_ref <- IRanges::subsetByOverlaps(x = file, ranges = reference_tidy, invert = TRUE)
        sample_not_in_ref_list <- c(sample_not_in_ref_list, sample_not_in_ref)
      }
      names(sample_not_in_ref_list) <- names(peaklist_tidy)
      # annotate unique peaks with chromHMM annotations
      sample_not_in_ref_chrmHMM <- plot_chrmHMM(peaklist = sample_not_in_ref_list,
                                                chrmHMM_annotation = chrHMM_list)
      print(sample_not_in_ref_chrmHMM)
      # save output 
      if(params$save_output){
        ggplot2::ggsave(filename = "sample_not_in_ref_chromHMM.png", plot = sample_not_in_ref_chrmHMM, device = "png", path = outfile_dir)
        unlink("sample_not_in_ref_chromHMM.png")
      }
  }
}
```

#### **Unique**: Reference peaks not in Sample peaks 

ChromHMM annotation of reference peaks not found in sample peaks. 


```{r, echo=FALSE, fig.width=9, warning=FALSE, message=FALSE}
if (params$chrmHMM_plot){
  if (is.null(params$reference)==FALSE){
      ref_not_in_sample_list <- list()
      # obtain unique peaks 
      for(file in peaklist_tidy){
        ref_not_in_sample <- IRanges::subsetByOverlaps(x = reference_tidy, ranges = file, invert = TRUE)
        ref_not_in_sample_list <- c(ref_not_in_sample_list, ref_not_in_sample)
      }
      names(ref_not_in_sample_list) <- names(peaklist_tidy)
      # annotate unique peaks with chromHMM annotations
      ref_not_in_sample_chrmHMM <- plot_chrmHMM(peaklist = ref_not_in_sample_list,
                                                chrmHMM_annotation = chrHMM_list)
      print(ref_not_in_sample_chrmHMM)
      # save output 
      if(params$save_output){
        ggplot2::ggsave(filename = "ref_not_in_sample_chromHMM.png", plot = ref_not_in_sample_chrmHMM, device = "png", path = outfile_dir)
        unlink("ref_not_in_sample_chromHMM.png")
      }
  }
}
```

### 3.2 ChIPseeker 

EpiCompare uses `annotatePeak` function in `ChIPseeker` package to annotate peaks with the nearest gene and 
genomic region where the peak is located. The peaks are annotated with [genes](http://bioconductor.org/packages/release/data/annotation/html/TxDb.Hsapiens.UCSC.hg19.knownGene.html) taken from the annotations of human genome hg19 provided by Bioconductor.

```{r, echo=FALSE, fig.width=9, warning=FALSE, message=FALSE}
# if set TRUE, generate chipseeker annotation plot 
if (params$chipseeker_plot){
  chipseeker_plot <- plot_ChIPseeker_annotation(peaklist = peaklist_tidy)
  
  print(chipseeker_plot)
  # save output 
  if(params$save_output){
    ggplot2::ggsave(filename = "chipseeker_annotation.png", plot = chipseeker_plot, device = "png", path = outfile_dir)
    unlink("chipseeker_annotation.png")
    }
}
```

### 3.3 Functional Enrichment Analysis {.tabset}

EpiCompare performs KEGG pathway and GO enrichment analysis using `clusterProfiler`. 
`annotatePeak` function in `ChIPseeker` package is first used to assign peaks to nearest genes 
and biological themes amongst the genes are identified using ontologies (KEGG and GO).
The peaks are annotated with [genes](http://bioconductor.org/packages/release/data/annotation/html/TxDb.Hsapiens.UCSC.hg19.knownGene.html) taken from the annotations of human genome hg19 provided by Bioconductor.


```{r, echo=FALSE, fig.width=9, warning=FALSE, message=FALSE}
fig_height <- 8
if (params$enrichment_plot){
  enrichment_plots <- plot_enrichment(peaklist = peaklist_tidy)
  KEGG_plot <- enrichment_plots[[1]]
  GO_plot <- enrichment_plots[[2]]
  kegg_terms <- length(unique(KEGG_plot$data$Description))
  go_terms <- length(unique(GO_plot$data$Description))
  if(kegg_terms > 20 | go_terms > 20){
    fig_height <- 10}
}
```

#### KEGG

```{r, echo=FALSE, fig.width= 10, fig.height = fig_height, warning=FALSE, message=FALSE}
if (params$enrichment_plot){
  print(KEGG_plot)
  # save output 
  if(params$save_output){
    ggplot2::ggsave(filename = "KEGG_analysis.png", plot = KEGG_plot, device = "png", path = outfile_dir)
    unlink("KEGG_analysis.png")
  }
}
```

#### GO 

```{r, echo=FALSE, fig.width=10,fig.height = fig_height, warning=FALSE, message=FALSE}
if (params$enrichment_plot){
  print(GO_plot)
  # save output 
  if(params$save_output){
    ggplot2::ggsave(filename = "GO_analysis.png", plot = GO_plot, device = "png", path = outfile_dir)
    unlink("GO_analysis.png")
  }
}

```
