#' Example ChIP-seq peak file
#'
#' Human H3K27ac peak file generated with ChIP-seq using K562 cell-line.
#' Raw peak file (.BED) was obtained from ENCODE project
#' (\url{https://www.encodeproject.org/files/ENCFF044JNJ/}).
#' The BED file was then imported as an GRanges object.
#' Peaks located on chromosome 1 were subsetted to reduce the dataset size.
#'
#' @source
#' The code to prepare the .Rda file from the raw peak file is:
#'
#' \code{
#' # dataset was directly downloaded from https://www.encodeproject.org/files/ENCFF044JNJ/
#' encode_H3K27ac <- ChIPseeker::readPeakFile("path", as = "GRanges")
#' encode_H3K27ac <- encode_H3K27ac[seqnames(encode_H3K27ac) == "chr1"]
#' colnames(GenomicRanges::mcols(encode_H3K27ac)) <- c("name","score","strand","signalValue","pValue","qValue","peak")
#' usethis::use_data(encode_H3K27ac, overwrite = TRUE)
#' }
#' @usage data("encode_H3K27ac")
"encode_H3K27ac"

#' Example CUT&Tag peak file
#'
#' Human H3K27ac peak file generated with CUT&Tag using K562 cell-line from
#' Kaya-Okur et al., (2019). Raw peak file (.BED) was obtained from GEO
#' (\url{https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR8383507}).
#' Peak calling was performed by Leyla Abbasova using MACS2.
#' The peak file was then imported as an GRanges object.
#' Peaks located on chromosome 1 were subsetted to reduce the dataset size.
#'
#' @source
#' The code to prepare the .Rda file from the raw peak file is:
#'
#' \code{
#' # sequences were directly downloaded from https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR8383507
#' # and peaks (BED file) were generated by Leyla Abbasova (Neurogenomics Lab, Imperial College London)
#' CnT_H3K27ac <- ChIPseeker::readPeakFile("path", as = "GRanges")
#' CnT_H3K27ac <- CnT_H3K27ac[seqnames(CnT_H3K27ac)== "chr1"]
#' colnames(GenomicRanges::mcols(CnT_H3K27ac)) <- c("name","score","strand","signalValue","pValue","qValue","peak")
#' usethis::use_data(CnT_H3K27ac)
#' }
#' @usage data("CnT_H3K27ac")
"CnT_H3K27ac"

#' Example CUT&Run peak file
#'
#' Human H3K27ac peak file generated with CUT&Run using K562 cell-line from
#' Meers et al., (2019).Raw peak file (.BED) was obtained from GEO
#' (\url{https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR8581604}).
#' Peak calling was performed by Leyla Abbasova using MACS2.
#' The peak file was then processed into GRanges object.
#' Peaks located on chromosome 1 were subsetted to reduce the dataset size.
#'
#' @source
#' The code to prepare the .Rda file from the raw peak file is:
#'
#' \code{
#' # sequences were directly downloaded from https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR8581604
#' # and peaks (BED file) were generated by Leyla Abbasova (Neurogenomics Lab, Imperial College London)
#' CnR_H3K27ac <- ChIPseeker::readPeakFile("path", as = "GRanges")
#' CnR_H3K27ac <- CnR_H3K27ac[seqnames(CnR_H3K27ac)== "chr1"]
#' colnames(GenomicRanges::mcols(CnR_H3K27ac)) <- c("name","score","strand","signalValue","pValue","qValue","peak")
#' usethis::use_data(CnR_H3K27ac, overwrite = TRUE)
#' }
#' @usage data("CnR_H3K27ac")
"CnR_H3K27ac"

#' Example Picard duplication metrics file 1
#'
#' Duplication metrics output of CUT&Tag H3K27ac file (sample accession: SRR8581604).
#' Raw sequences were aligned to hg19 genome and Picard was performed by Leyla Abbasova.
#' The duplication summary output generated by Picard was processed to reduce the size of data.
#'
#' @source
#' The code to prepare the .Rda file is:
#'
#' \code{
#' picard <- read.table("path/to/picard/duplication/output", header = TRUE, fill = TRUE)
#' CnT_H3K27ac_picard <- picard[1,]
#' usethis::use_data(CnT_H3K27ac_picard, overwrite = TRUE)
#' }
#' @usage data("CnT_H3K27ac_picard")
"CnT_H3K27ac_picard"

#' Example Picard duplication metrics file 2
#'
#' Duplication metrics output on CUT&Run H3K27ac file (sample accession: SRR8581604).
#' Raw sequences were aligned to hg19 genome and after, Picard was performed by Leyla Abbasova.
#' The duplication summary output generated by Picard was processed to reduce the size of data.
#'
#' @source
#' The code to prepare the .Rda file is:
#'
#' \code{
#' picard <- read.table("path/to/picard/duplication/output", header = TRUE, fill = TRUE)
#' CnR_H3K27ac_picard <- picard[1,]
#' usethis::use_data(CnR_H3K27ac_picard, overwrite = TRUE)
#' }
#' @usage data("CnR_H3K27ac_picard")
"CnR_H3K27ac_picard"

#' Human genome hg19 blacklisted regions
#'
#' Obtained from \url{https://www.encodeproject.org/files/ENCFF001TDO/}.
#' The ENCODE blacklist includes regions of the genome that have anomalous
#' and/or unstructured signals independent of the cell-line or experiment.
#' Removal of ENCODE blacklist is recommended for quality measure.
#'
#' @source
#' The code to prepare the .Rda file is:
#'
#' \code{
#' # blacklisted regions were directly downloaded
#' # from \url{https://www.encodeproject.org/files/ENCFF001TDO/}
#' hg19_blacklist <- ChIPseeker::readPeakFile(file.path(path), as = "GRanges")
#' usethis::use_data(hg19_blacklist, overwrite = TRUE)
#' }
#' @usage data("hg19_blacklist")
"hg19_blacklist"

#' Human genome hg38 blacklisted regions
#'
#' Obtained from \url{https://www.encodeproject.org/files/ENCFF356LFX/}.
#' The ENCODE blacklist includes regions of the genome that have anomalous and/
#' or unstructured signals independent of the cell-line or experiment. Removal
#' of ENCODE blacklist is recommended for quality measure.
#'
#' @source
#' The code to prepare the .Rda file is:
#'
#' \code{
#' ## blacklisted regions were directly downloaded
#' ## from \url{https://www.encodeproject.org/files/ENCFF356LFX/}
#' hg38_blacklist <- ChIPseeker::readPeakFile(file.path(path), as = "GRanges")
#' usethis::use_data(hg38_blacklist, overwrite = TRUE)
#' }
#' @usage data("hg38_blacklist")
"hg38_blacklist"

